File: easy/257_binary_tree_paths.rb
# 257. Binary Tree Paths
# Easy
#
# Given the root of a binary tree, return all root-to-leaf paths in any order.
#
# A leaf is a node with no children.

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

def binary_tree_paths(root)
  return nil if root.nil?

  helper(root, [])
end

def helper(node, path)
  path = path.empty? ? [node.val.to_s] : path.map { |p| "#{p}->#{node.val}" }

  if !node.left.nil? || !node.right.nil?
    left = node.left.nil? ? [] : helper(node.left, path)
    right = node.right.nil? ? [] : helper(node.right, path)
    left.concat(right)
  else
    path
  end
end

File: python/easy/257_binary_tree_paths.py
"""
257. Binary Tree Paths
Easy

Given the root of a binary tree, return all root-to-leaf paths in any order.

A leaf is a node with no children.
"""


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if root is None:
            return []

        return helper(root, [])


def helper(root, paths):
    if len(paths) == 0:
        new_paths = [str(root.val)]
    else:
        new_paths = [p + f"->{root.val}" for p in paths]

    if root.left is None and root.right is None:
        return new_paths

    left = []
    if root.left is not None:
        left = helper(root.left, new_paths)

    right = []
    if root.right is not None:
        right = helper(root.right, new_paths)

    return left + right
